Image block: Revise lightbox UI to remove 'behaviors'

artemiomorales opened the pull request:
## What?
Now that the concept of [behaviors](https://github.com/WordPress/gutenberg/issues/50029) is no longer shipping in the near term, this PR removes it from the `theme.json`, the Global Styles and the Image block attributes.

## Why?
We want to provide a lightbox feature and get feedback on its implementation, but we also want to do so in a way that's minimal and easy to understand. Addresses #53403 

## How?
Removing all the references to `behaviors`. 

_Note regarding backwards compatibility: We've added backwards compatibility [in the following PR](https://github.com/WordPress/gutenberg/pull/54071), which is meant to be merged at the same time as this one so there's no gap in support for users using the old syntax we're removing in this PR._

## Testing Instructions
1. Open a post or page.
2. Insert an Image block.
3. Make sure that no "Behaviors" panel is preset inside of the "Advanced" panel of the Inspector Controls.
4. Save the post/page
5. Go to the frontend
6. Make sure that the page loads correctly and there is no Lightbox present.

artemiomorales committed:
Begin removing theme.json dependency in block UI

michalczaplinski committed:
Remove the useHasBehaviorsPanel hook

michalczaplinski commented on packages/block-editor/src/components/global-styles/hooks.js:
```diff
@@ -546,27 +546,3 @@ export function __experimentalUseGlobalBehaviors( blockName, source = 'all' ) {
 	return { behavior, inheritedBehaviors: result, setBehavior };
 }
-
-export function __experimentalUseHasBehaviorsPanel(
-	settings,
-	name,
-	{ blockSupportOnly = false } = {}
-) {
-	if ( ! settings?.behaviors ) {
-		return false;
-	}
-
-	// If every behavior is disabled on block supports, do not show the behaviors inspector control.
-	const hasSomeBlockSupport = Object.keys( settings?.behaviors ).some(
-		( key ) => hasBlockSupport( name, `behaviors.${ key }` )
-	);
-
-	if ( blockSupportOnly ) {
-		return hasSomeBlockSupport;
-	}
-
-	// If every behavior is disabled, do not show the behaviors inspector control.
-	return Object.values( settings?.behaviors ).some(
-		( value ) => value === true && hasSomeBlockSupport
-	);
-}
```
If I'm not mistaken, we'll always want to show the "Behaviors" Panel (under a different name probably) in the UI now that the `theme.json` support is disabled. So this hook is not necessary anymore.

michalczaplinski commented on lib/block-supports/behaviors.php:
```diff
@@ -50,9 +50,10 @@ function gutenberg_render_behaviors_support_lightbox( $block_content, $block ) {
-		$theme_data = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data()->get_data();
-		if ( isset( $theme_data['behaviors']['blocks'][ $block['blockName'] ]['lightbox'] ) ) {
-			$lightbox_settings = $theme_data['behaviors']['blocks'][ $block['blockName'] ]['lightbox'];
+		$user_data = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data()->get_data();
```
Should we not use the `::get_user_data()` method here after all?

artemiomorales replied:
Whoops, I did this wrong. Thanks for the catch!

artemiomorales committed:
Fix  declaration to actually retrieve from user data

michalczaplinski committed:
Restructured use of global behaviors

Simplified the `__experimentalUseGlobalBehaviors` function by removing the 'source' parameter. Now, the function directly uses 'userConfig' for getting raw data and 'mergedConfig' for variable value.

michalczaplinski commented:
I've pushed some last changes. There are a couple of things that are not working and we need final decision on how to structure the options. I've explained the changes here:

<div>
    <a href="https://www.loom.com/share/7c970f1314714a55b1ee46ff5afaa3bc">
      <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/7c970f1314714a55b1ee46ff5afaa3bc-with-play.gif">
    </a>
  </div>

michalczaplinski committed:
More removal of behaviors.

michalczaplinski committed:
Remove the reference to behaviors in Global styles and first iteration of updates to the lightbox UI

michalczaplinski committed:
Remove behaviors altogether and everywhere

michalczaplinski committed:
Fix linter & sniffer PHP errors

michalczaplinski committed:
Adjust schema properties count assertion

Previously, the test was checking whether schema properties array had exactly 10 elements
We're now checking for exactly 9 elements instead.

michalczaplinski commented on packages/block-library/src/image/deprecated.js:
```diff
@@ -632,9 +632,6 @@ const v6 = {
 	},
 	supports: {
 		anchor: true,
-		behaviors: {
-			lightbox: true,
-		},
```
This one I'm not 100% sure about. 

I think that we can safely remove the reference to behaviors in those deprecations because if a user does have a block with the `behaviors: { ...}` attribute for some reason, it should just be ignored. 

Could you confirm that @gziolo ?

gziolo replied:
Block supports can be safely removed. The only question is what will happen when there is a value set for the `behaviors` block attribute set in the saved HTML as part of the comment metadata. As far as I can tell, it was: `behaviors.lightbox` (based on `$attributes['behaviors']['lightbox']`) so it would be nice to transfer the value to whatever else you plan to use moving forward. 

One way to test whether renaming the attribute is safe, would be pasting the old saved Image block with the lightbox enabled locally and checking whether the block editor won't trigger the block validation error. You can also simulate that by adding a new block fixture in integration tests as explained in https://github.com/WordPress/gutenberg/blob/trunk/test/integration/fixtures/blocks/README.md.

ajlende replied:
We should probably add a block deprecation for this either way, right? If there is no validation error, you should be able to set an `isElegible` method on the deprecation based on the presence of the top-level `behaviors` attribute. And then the `behaviors` attribute can be migrated to the `lightbox` attribute with the `migrate` method.

This would save having to add the deprecation warning in https://github.com/WordPress/gutenberg/pull/54071/files#diff-a14b61c874095649b236b78413e805b4541277b24eb71ce350a17d9f2b6396b1R765-R779.

artemiomorales replied:
> One way to test whether renaming the attribute is safe, would be pasting the old saved Image block with the lightbox enabled locally and checking whether the block editor won't trigger the block validation error.

I've tested this by adding the JSON manually to the block and there's no error, so it looks like renaming the attribute is safe.

> We should probably add a block deprecation for this either way, right?

@ajlende This sounds right.

As per [this comment](https://github.com/WordPress/gutenberg/pull/54071#discussion_r1318511575), it looks like we can handle the migration in a separate PR later, right? 

We'll be offering backwards compatibility for 2 Gutenberg releases. As this PR is already pretty big, I think it makes sense to just push this forward and handle the deprecation separately.

ajlende replied:
> As per https://github.com/WordPress/gutenberg/pull/54071#discussion_r1318511575, it looks like we can handle the migration in a separate PR later, right?
>
> We'll be offering backwards compatibility for 2 Gutenberg releases. As this PR is already pretty big, I think it makes sense to just push this forward and handle the deprecation separately.

[Responded below.](https://github.com/WordPress/gutenberg/pull/53851#issuecomment-1712082579)

michalczaplinski committed:
Add the lightbox attribute

michalczaplinski commented on packages/block-library/src/image/index.php:
```diff
@@ -33,17 +33,10 @@ function render_block_core_image( $attributes, $content, $block ) {
-		} else {
-			$lightbox_settings = null;
-		}
+	if ( isset( $attributes['lightbox'] ) ) {
+		$lightbox_settings = $attributes['lightbox'];
```
I'm assuming that this will be the structure of the "new" attribute that controls the lightbox.

michalczaplinski commented on packages/block-library/src/image/index.php:
```diff
@@ -71,6 +64,202 @@ function render_block_core_image( $attributes, $content, $block ) {
+	return str_replace( '</figure>', $lightbox_html . '</figure>', $body_content );
+}
+
+// TODO: We should not be adding a separate filter but rather move the
+// the lightbox rendering to the `render_block_core_image` function.
+
+// Use priority 15 to run this hook after other hooks/plugins.
+// They could use the `render_block_{$this->name}` filter to modify the markup.
+add_filter( 'render_block_core/image', 'block_core_image_render_lightbox', 15, 2 );
```
Noting the TODO for the subsequent PR that adds the revised Lightbox implementation.

michalczaplinski committed:
Remove unnecessary space

michalczaplinski commented on packages/block-library/src/image/index.php:
```diff
@@ -71,6 +64,202 @@ function render_block_core_image( $attributes, $content, $block ) {
+function block_core_image_render_lightbox( $block_content, $block ) {
+	$link_destination = isset( $block['attrs']['linkDestination'] ) ? $block['attrs']['linkDestination'] : 'none';
+
+	// Get the lightbox setting from the block attributes.
+	if ( isset( $block['attrs']['lightbox'] ) ) {
+		$lightbox_settings = $block['attrs']['lightbox'];
+	}
```
As in the previous comment, I'm also assuming that this will be the structure of the "new" attribute that controls the lightbox.

michalczaplinski commented on test/e2e/specs/editor/blocks/image.spec.js:
```diff
@@ -728,7 +728,7 @@ test.describe( 'Image', () => {
 } );
 
-test.describe( 'Image - interactivity', () => {
+test.describe.skip( 'Image - interactivity', () => {
```
I'm going to skip this test for now because we'll have to adjust the tests in the subsequent PR to use the "new" Lightbox UI.

ramonjd replied:
Would be it okay to add a comment here for people like me who forget stuff (like looking at Github annotations) 😄 ?

artemiomorales replied:
Done 👍

michalczaplinski commented on test/e2e/specs/editor/blocks/image.spec.js:
```diff
@@ -927,104 +927,6 @@ test.describe( 'Image - interactivity', () => {
 				await expect( lightbox ).toBeHidden();
 			} );
-
-			test( 'fade animation', async ( {
```
There's no fade animation anymore (only "expand") so we can delete this test.

michalczaplinski commented on test/e2e/specs/editor/blocks/image.spec.js:
```diff
@@ -1117,108 +1019,6 @@ test.describe( 'Image - interactivity', () => {
 			).not.toBeInViewport();
 		} );
 
-		test.describe( 'Animation Select visibility', () => {
```
Removing this test as well, because there is no "animation selector" anymore. The Lightbox only has the "zoom" effect (animation) - it's the only one.

michalczaplinski commented on lib/theme.json:
```diff
@@ -1,15 +1,5 @@
 {
 	"version": 2,
-	"behaviors": {
-		"blocks": {
-			"core/image": {
-				"lightbox": {
-					"enabled": false,
-					"animation": ""
-				}
-			}
-		}
-	},
```
Can this just be removed or does it need a deprecation of some sort? Behaviors were [shipped](https://github.com/WordPress/gutenberg/releases/tag/v16.4.0) in GB 16.4 so never included in a WP release.

cc @ellatrix @gziolo @youknowriad

gziolo replied:
@artemiomorales and I looked at whether this is used in production, and it looks like you can change that setting in the UI, which most likely gets reflected in what gets saved in the database. While it isn't mandatory for WordPress core, it would still be nice to provide a migration path for sites that have the latest Gutenberg plugin enabled. It shouldn't be that difficult to code something in the plugin to read `behaviors` object from the site and transfer the value to the new structure.

I guess the same applies to the block attribute that changes from `behaviors.lightbox` to `lightbox` in the Image block.

oandregal replied:
I'm not fully caught up on what's next for behaviors, but it seems that themes will no longer be able to tweak the lightbox feature (enable/disable, set animation) after this removal.

If we still plan to allow them to tweak it, just in a different place, we could add some code around [this line](https://github.com/WordPress/gutenberg/blob/trunk/lib/class-wp-theme-json-gutenberg.php#L617) to make the migration easier for them. I'll be AFK in the next few days, but wanted to share in case it's useful.

michalczaplinski replied:
In the followup PR: https://github.com/WordPress/gutenberg/pull/54071 we've added backwards-compatibility for users that might have been using the previous `behaviors.lightbox` syntax. 

In that PR (which is based on the current one) we're also now introducing a `lightbox.allowEditing` setting which allows themes to enable/disable the lightbox.

michalczaplinski committed:
Merge branch 'trunk' into update/revise-lightbox-ui

ramonjd commented:
I've taken this PR for a spin, and checked the global styles/revisions rest API changes and associated tests. From that perspective, it looks good to me. 

On the backwards compatibility subject - I've heard it more than once that the plugin itself supports the previous two versions only, when those changes do not impact on Core. 

Here's that noted in a comment: https://github.com/WordPress/gutenberg/pull/47475#issuecomment-1409560457

But I can't yet find anything official. I think it's because there is nothing official. 

Was this published as an experimental API? The use of `__experimental` suggest yes. Will it affect anyone who updates to the latest WordPress version? Probably not, right? These are some things we could ask ourselves in two versions' time.

I guess we should be always wary when it's a public API. It was a topic discussed at the WordPress community summit this year. See: https://make.wordpress.org/summit/2023/08/23/community-summit-discussion-notes-addressing-backwards-compatibility-in-gutenberg/

Related issue:

- https://github.com/WordPress/gutenberg/issues/41265

michalczaplinski commented:
Thanks a lot for the review, Ramon! 

> Was this published as an experimental API? The use of __experimental suggest yes.
The API itself can be considered experimental, but since it's exposed in the UI, there's not really a way for users to know if what they are enabling is experimental or not. It was not explicitly described as "experimental" in the UI. 

It was never shipped in Core, though, so I guess deprecation is the way to go!

Just to let you know, there is a follow-up PR to this one: https://github.com/WordPress/gutenberg/pull/54071/
There, we add the updated version of the Lightbox (without behaviors).

artemiomorales committed:
Add clarifying comment regarding skipped tests

michalczaplinski committed:
Do not remove `behaviors` attribute from Image block's block.json.
Behaviors are deprecated for 2 more releases and will be removed then.

michalczaplinski committed:
Revert "Do not remove `behaviors` attribute from Image block's block.json."

This reverts commit cabe31b902440b227bfa0bb7f670c2942a1aeba7.

ramonjd reviewed (approved):
I think the state of this PR is good, thanks for sticking with it. 

If @gziolo or @ajlende don't have any further concerns, then it LGTM.

I understand this needs to be merged before https://github.com/WordPress/gutenberg/pull/54071, so I suppose we could address follow up concerns in that PR?

ajlende commented:
I tested with the following block content and theme.json modification.

```html
<!-- wp:heading -->
<h2 class="wp-block-heading">Global</h2>
<!-- /wp:heading -->

<!-- wp:image {"id":68,"sizeSlug":"large","linkDestination":"none"} -->
<figure class="wp-block-image size-large"><img src="http://wp.local/wp-content/uploads/2023/02/27982606245_0a702a9b26_o-1024x576.jpg" alt="" class="wp-image-68"/></figure>
<!-- /wp:image -->

<!-- wp:heading -->
<h2 class="wp-block-heading">Block</h2>
<!-- /wp:heading -->

<!-- wp:image {"id":68,"sizeSlug":"large","linkDestination":"none","behaviors":{"lightbox":{"enabled":true,"animation":"fade"}}} -->
<figure class="wp-block-image size-large"><img src="http://wp.local/wp-content/uploads/2023/02/27982606245_0a702a9b26_o-1024x576.jpg" alt="" class="wp-image-68"/></figure>
<!-- /wp:image -->
```

```jsonc
{
	// ...
	"behaviors": {
		"blocks": {
			"core/image": {
				"lightbox": {
					"enabled": true
				}
			}
		}
	},
	// ...
}
```

This is what I see.

| Before | After |
| ------ | ----- |
| ![behaviors-deprecation-lightbox-before](https://github.com/WordPress/gutenberg/assets/5129775/1b8af93d-2514-4a8a-8ee0-b0ae587a28f4) | ![behaviors-deprecation-lightbox-after](https://github.com/WordPress/gutenberg/assets/5129775/e2769555-79e4-4cfa-a67f-6e30a05274cc) |
| ![behaviors-deprecation-lightbox-global-before](https://github.com/WordPress/gutenberg/assets/5129775/4200e1b1-b776-4d7b-8b65-405408d79af0) | ![behaviors-deprecation-lightbox-global-after](https://github.com/WordPress/gutenberg/assets/5129775/b6e57eb9-fe8a-459a-aba6-07de898208ae) |
| ![behaviors-deprecation-lightbox-block-before](https://github.com/WordPress/gutenberg/assets/5129775/d3d5a0ce-a49c-48fb-9e4f-f10d5e99d97e) | ![behaviors-deprecation-lightbox-block-after](https://github.com/WordPress/gutenberg/assets/5129775/d58fc5f6-0103-4014-8a62-65209efff384) |

> As per [this comment](https://github.com/WordPress/gutenberg/pull/54071#discussion_r1318511575), it looks like we can handle the migration in a separate PR later, right?

@artemiomorales The lightbox support wasn't hidden behind an experimental flag or anything in the UI, so folks using the plugin who have been using lightbox will see this as breaking their site if you don't have the migration. So I would recommend getting the migration handled before merging this PR so we can be _sure_ they make it into the same GB release.

> We'll be offering backwards compatibility for 2 Gutenberg releases. As this PR is already pretty big, I think it makes sense to just push this forward and handle the deprecation separately.

As this PR is right now, there is no backwards compatibility. The functionality has simply been removed.

We have systems for migrating both block attributes and global styles settings that should be used.

For block attributes, you just need to migrate the attributes to their new location with a block deprecation. Since the block supports code is removed, you'll have to replace the supports in the block deprecations with the `behaviors` attribute in `attributes`. Then you can use the `isEligible` method to trigger the migration.

For global styles (theme.json), we also have a migration system in class-wp-theme-json-schema.php. You can write a migration function to move the setting where you wish.

The block supports part is the only part where you'll need to properly add a deprecation. That can be done simply by adding a [`_deprecated_function()`](https://developer.wordpress.org/reference/functions/_deprecated_function/) call to the functions and removing only the block support registration.

Any public function you want to remove can't be deleted if you're providing backwards compatibility. Instead, you add one of our deprecated functions that will print a warning. Then in two versions, the function can finally be removed.

michalczaplinski committed:
Merge branch 'trunk' into update/revise-lightbox-ui

artemiomorales committed:
Revert deletion of behaviors.php

michalczaplinski committed:
Merge branch 'trunk' into update/revise-lightbox-ui

artemiomorales committed:
Image block: UI updates for the image lightbox (#54071)

* Add initial implementation of image settings panel

* Remove unnecessary code and fix reset functionality

* Add UI to image block inspector

* Add `lightbox` to the valid `theme.json` settings

* Fix a bug with image selector and integrate with global styles

* Update `theme.json` schema

* Added the `@since`annotation

* Refactor image settings panel and screen block

Simplified the ImageSettingsPanel and ScreenBlock components. More specific changes:
 - Removed `name` and `settings` from the ImageSettingsPanel
 - Use `userSettings` instead of `settings` in the ImageSettingsPanel
 - Modified `onChangeLightbox` logic, it now takes a new setting instead of a boolean and directly passes to `onChange`
 - Updated the ScreenBlock component to account for this refactor

* Add showUI option to lightbox settings

A new option has been added to the lightbox settings in the Theme JSON reference guide and Gutenberg class. This `showUI` option allows users to toggle whether the Lightbox UI is displayed in the block editor or not. Also, updated the JSON schema accordingly to reflect these changes in theme.json files.

* Add defaults for the `lightbox` to the GB `theme.json`

* Change the falsy checks in image.js

* Add filters; add legacy support for behaviors syntax

I moved the logic to determine whether the lightbox should display
or not to two render_block_data filters.

One of these filters is inside of the index.php
so that itc can exist in WP core, the other inside of blocks.php
in order to offer legacy support for the Behaviors syntax in
the Gutenberg plugin.

Using the render_block_data instead of render_block allows us to
store a 'lightboxEnabled' value on the block, which we can use to
determine whether the lightbox should be rendered in these two
separate locations relatively cleanly without needing to touch
the markup.

I added behaviors back to the valid top-level keys so that we can
read it to offer legacy support.

Lastly, I set the lightbox.enabled attribute to NULL by default
so that we can determine whether the Behaviors syntax should override
it or not.

* Fix linter errors & add more expansive comments.

* If no value is set for the lightbox in the Global Styles, then the block editor UI should inherit the value from `theme.json`.

Likewise, if no value is set in the block attributes, the block editor UI should inherit the value from the Global Styles of the Image.

* Rename `showUI` to `allowEditing`

* Fix the `theme.json` schema

* Use the globalPath for the settings on the PHP side

* Add backwards support for enabling fade animation via the legacy syntax

* Fix PHP linter errors

* Fix error when checking lightbox['enabled'] value in global settings

* Empty commit

* Remove the default `false` value for `lightbox.enabled`attribute.

* Add deprecation notice for 'behaviors' in image block

I needed to add 'behaviors' back to the block.json attributes
in order to read them on the JavaScript side in the editor
to fire the deprecation notice.

* Add deprecation for  attribute in image block

* Remove obsolete code now that block deprecation is in place

* Add support for 'lightbox: true' syntax

* Fix lightbox 'checked' attribute being read improperly

* Add conditional display for settings panel at image block level

* Fix an error with the theme.json schema.

* Update docs with `npm run build:docs`

* Copy `class-wp-theme-json-schema.php` from core
 into `class-wp-theme-json-schema-gutenberg.php`

* Add a theme.json migration to v3 away from behaviors and to a new, simpler syntax used by the lightbox.

* Remove the `null` value from lightbox.enabled in `lib/theme.json`

* Remove `behaviors` from VALID_TOP_LEVEL_KEYS & VALID_SETTINGS

* Revise backwards compatibility for behaviors; add deprecation to block_supports

* Remove outdated comment

* Update outdated comment

* Update comment and shuffle the lines so the diff is easier on the eyes.

* Update comment to explain why we use userSettings in image-settings-panel.js

* Remove support for legacy fade configuration

* Resolve lint error

* Add clarifying comment regarding lightbox markup

* Rename the migrate function to reflect that it's not a v3 migration

* Add a `@since` in `gutenberg_should_render_lightbox` docblock

* Add support for reading top-level 'lightbox' setting in editor

By default, we read the lightbox settings underneath the 'core/image'
in theme.json; however, the 'enabled' property there is undefined by default,
which means it should be possible to declare a top-level setting for the lightbox
that overrides an undefined block-level setting.

While this appeared to be working on the PHP side, the UI wasn't accurately
reflecting this inheritance structure, so this commit fixes that.

Users should now be able to define a top-level lightbox setting as either
'lightbox: true' or 'lightbox: { enabled: true }' that will be used
if the block-level lightbox setting for 'enabled' is undefined.

* Revert "Add support for reading top-level 'lightbox' setting in editor"

This reverts commit 2f5f122e5e2e505d4ccacffc2957df15cd63d628.

* Add correct deprecation mentioning the Gutenberg version

* Move 'allowEditing' to top-level settings

* Fix top-level lightbox setting not being read properly

* Fix 'false' values in theme.json not being stored in settings object

* Fix error wherein 'undefined' was being passed to input component

* Fix bug wherein lightbox UI would disappear if user value was set

* Remove inheritance when determining whether to enable lightbox

Rather than trying to check if the 'enabled' has been set or not
and falling back to other levels of the theme.json inheritance
structure, I decided to just read and use the settings as defined
in theme.json. This is the expected behavior in Gutenberg from
what I understand and has less edge cases.

* Update comment

* Update whitespace in theme.json

Co-authored-by: Alex Lende <alex+github.com@lende.xyz>

* Add docblocks to clarify that `class-wp-theme-json-schema-gutenberg.php` is only put in GB as a temporary migration.

* Add comments to clarify that behaviors.php is temporarily added to GB an will be removed in a future version.

* Added integration fixtures for the lightbox

* Fix incorrect reading of global lightbox settings

* Clarify the comment getting the global lightbox settings

* Fix PHPCS parenthesis error 🤦‍♂️

* Move lightbox settings to the block level

* Remove support for shorthand

* Remove 'lightbox' from hooks.js and add `.enabled` & `.allowEditing` to class-wp-theme-json-gutenberg.php

---------

Co-authored-by: Michal Czaplinski <mmczaplinski@gmail.com>
Co-authored-by: Alex Lende <alex+github.com@lende.xyz>

artemiomorales committed:
Revert "Image block: UI updates for the image lightbox (#54071)"

This reverts commit 1b3a48049a1210169bb81fac7f9970c092be991d.

michalczaplinski commented:
The backwards compatibility concerns have been addressed in https://github.com/WordPress/gutenberg/pull/54509 👍
